csharp/src/Google.Protobuf/Reflection/Dynamic/DynamicMessage.cs

```diff
@@ -0,0 +1,383 @@
+            {
+                fields.SetField(fd, value);
+            }
+
+            /// <summary>
+            /// Used to add repeated field
+            /// </summary>
+            /// <param name="fd"></param>
+            /// <param name="v"></param>
+            public void AddRepeatedField(FieldDescriptor fd, object v)
+            {
+                fields.AddRepeatedField(fd, v);
+            }
+
+            private DynamicMessage BuildPartial()
+            {
+                if (fields == null)
+                {
+                    throw new InvalidOperationException("Build() has already been called on this Builder.");
+                }
+                DynamicMessage result = new DynamicMessage(type, fields);
+                fields = null;
+                return result;
+            }
+
+            /// <summary>
+            /// Reads every field and sets it in the FieldSet instance
+            /// </summary>
+            /// <param name="input"></param>
+            /// <exception cref="Exception"></exception>
+            public void MergeFrom(CodedInputStream input)
+            {
+                uint tag;
+                while ((tag = input.ReadTag()) != 0)
+                {
+                    int fieldNumber = WireFormat.GetTagFieldNumber(tag);
+                    FieldDescriptor fd = type.FindFieldByNumber(fieldNumber);
+
+                    if (fd == null)
+                    {
+                        throw new Exception($"Field descriptor not found for fieldNumber: {fieldNumber}");
+                    }
+                    if (fd.FieldType == FieldType.Message)
+                    {
+                        MergeFromMessageTypeField(input, fd);
+                    }
+                    else
+                    {
+                        MergeFromPrimitiveTypeField(input, tag, fd);
+                    }
+                }
+            }
+
+            private void MergeFromPrimitiveTypeField(CodedInputStream input, uint tag, FieldDescriptor fd)
+            {
+                if (fd.ToProto().Label != FieldDescriptorProto.Types.Label.Repeated)
+                {
+                    object value = ReadField(fd.FieldType, input);
+                    fields.SetField(fd, value);
+                }
+                else
+                {
+                    IEnumerator enumerator = GetFieldCodec(tag, fd.FieldType, input);
+                    while (enumerator.MoveNext())
+                    {
+                        fields.AddRepeatedField(fd, enumerator.Current);
+                    }
+                }
+
+            }
+
+            private void MergeFromMessageTypeField(CodedInputStream input, FieldDescriptor fd)
+            {
+                Builder value = NewBuilder(fd.MessageType);
+                input.ReadMessage(value);
+                DynamicMessage val = value.Build();
+                if (fd.ToProto().Label != FieldDescriptorProto.Types.Label.Repeated)
```

Comment:
Nit: curly braces around blocks for `if` and `else` sides of this.

Reply:
sure done.